src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -22,121 +22,194 @@
+    private boolean mbeanRegistered = false;
 
     public final IEndpointSnitch subsnitch;
 
-    private volatile ScheduledFuture<?> updateSchedular;
-    private volatile ScheduledFuture<?> resetSchedular;
+    private volatile ScheduledFuture<?> updateScheduler;
+    private volatile ScheduledFuture<?> latencyProbeScheduler;
 
-    private final Runnable update;
-    private final Runnable reset;
+    private final Runnable update = this::updateScores;
+    private final Runnable latencyProbe = this::maybeSendLatencyProbe;
 
     public DynamicEndpointSnitch(IEndpointSnitch snitch)
     {
         this(snitch, null);
     }
 
-    public DynamicEndpointSnitch(IEndpointSnitch snitch, String instance)
+    protected DynamicEndpointSnitch(IEndpointSnitch snitch, String instance)
     {
         mbeanName = "org.apache.cassandra.db:type=DynamicEndpointSnitch";
         if (instance != null)
             mbeanName += ",instance=" + instance;
         subsnitch = snitch;
-        update = new Runnable()
-        {
-            public void run()
-            {
-                updateScores();
-            }
-        };
-        reset = new Runnable()
-        {
-            public void run()
-            {
-                // we do this so that a host considered bad has a chance to recover, otherwise would we never try
-                // to read from it, which would cause its score to never change
-                reset();
-            }
-        };
 
         if (DatabaseDescriptor.isDaemonInitialized())
         {
-            updateSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(update, dynamicUpdateInterval, dynamicUpdateInterval, TimeUnit.MILLISECONDS);
-            resetSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(reset, dynamicResetInterval, dynamicResetInterval, TimeUnit.MILLISECONDS);
+            applyConfigChanges();
             registerMBean();
         }
+        nextProbeGenerationTime = System.nanoTime();
+    }
+
+    /**
+     * Records a latency. This MUST be cheap as it is called in the fast path
+     */
+    abstract public void receiveTiming(InetAddressAndPort address, long latency, LatencyMeasurementType measurementType);
+
+    /**
+     * Called periodically from {@link DynamicEndpointSnitch#updateScores()}. This method returns a new score map
+     * that will be assigned to the {@link DynamicEndpointSnitch#scores} map.
+     * This is generally more expensive than {@link DynamicEndpointSnitch#receiveTiming}
+     */
+    abstract public Map<InetAddressAndPort, Double> calculateScores();
+
+    /**
+     * Signals that we actually tried to rank vs this host. This way any latency probing can just focus
+     * on hosts that this peer would potentially talk to.
+     * @param address
+     */
+    abstract protected void markRequested(InetAddressAndPort address);
```

Comment:
It was there because I had the wrapped metrics (reservoir/EMA) in the subclasses (as they don't both e.g. conform to some interface that I can then compose on top of). I'm trying to implement your suggestion of moving samples back to base class and along the way I wouldn't need this anymore.